File: Sources/goserver/internal/wave/wave_room_code.go
package wave

import (
	"fmt"
	"math/rand/v2"
	"regexp"
	"strings"
)

type WaveRoomCode string

func generateRandomServerIdentifier() string {
	chars := make([]string, 3)
	for i := range chars {
		chars[i] = fmt.Sprintf("%x", rand.IntN(16))
	}

	return strings.Join(chars, "")
}

func generateRandomMeaninglessIdentifier() string {
	chars := make([]string, 6)
	for i := range chars {
		chars[i] = string(rune('a' + rand.IntN(26)))
	}

	return strings.Join(chars, "")
}

func GenerateRandomWaveRoomCode() WaveRoomCode {
	serverPart := generateRandomServerIdentifier()
	meaninglessPart := generateRandomMeaninglessIdentifier()

	return WaveRoomCode(fmt.Sprintf("%s-%s", serverPart, meaninglessPart))
}

func IsWaveRoomCode(maybeCode string) bool {
	if len(maybeCode) != 10 {
		return false
	}
	
	if maybeCode[3] != '-' {
		return false
	}

	parts := strings.Split(maybeCode, "-")
	if len(parts) != 2 {
		return false
	}

	prefix := parts[0]
	suffix := parts[1]

	hexMatch, _ := regexp.MatchString("^[0-9a-f]{3}$", prefix)
	if !hexMatch {
		return false
	}

	alphaMatch, _ := regexp.MatchString("^[a-z]{6}$", suffix)
	if !alphaMatch {
		return false
	}

	return true
}

File: Sources/LegacyClient/Private/Sources/Wave/WaveRoomCode.ts
type ServerId = string & { length: 3 };
type MeaninglessId = string & { length: 6 };

// XXX-XXXXXX
export type WaveRoomCode = `${ServerId}-${MeaninglessId}`;

function generateRandomServerIdentifier(): ServerId {
    return [...Array(3)]
        .map(() => Math.floor(Math.random() * 16).toString(16))
        .join("") as ServerId;
}

function generateRandomMeaninglessIdentifier(): MeaninglessId {
    return [...Array(6)]
        .map(() => String.fromCharCode(97 + Math.floor(26 * Math.random())))
        .join("") as MeaninglessId;
}

export function generateRandomWaveRoomCode(): WaveRoomCode {
    const serverPart = generateRandomServerIdentifier();
    const meaninglessPart = generateRandomMeaninglessIdentifier();

    return `${serverPart}-${meaninglessPart}` satisfies WaveRoomCode;
}

export function isWaveRoomCode(maybeCode: string): maybeCode is WaveRoomCode {
    if (maybeCode.length !== 10) return false;
    if (maybeCode[3] !== "-") return false;

    const [prefix, suffix] = maybeCode.split("-");

    // Check if prefix is valid hex (3 characters)
    if (prefix.length !== 3 || !/^[0-9a-f]{3}$/.test(prefix)) return false;

    // Check if suffix is valid lowercase letters (6 characters)
    if (suffix.length !== 6 || !/^[a-z]{6}$/.test(suffix)) return false;

    return true;
}
